maro/simulator/scenarios/finance/common.py

```diff
@@ -0,0 +1,180 @@
+            is_trade_trigger: bool = True):
+        self.sub_engine_name = sub_engine_name
+        self.item_index = item_index
+        self.trade_number = trade_number
+        self.tick = tick
+        self.price_per_item = price_per_item
+        self.tax = tax
+        self.is_trade_accept = is_trade_accept
+        self.is_trade_trigger = is_trade_trigger
+        self.action_id = action_id
+
+    @property
+    def total_cost(self):
+        return self.trade_number * self.price_per_item + self.tax
+
+    def __repr__(self):
+        return f"< trade sub-engine: {self.sub_engine_name} item: {self.item_index} \
+            tick: {self.tick} number: {self.trade_number} price: {self.price_per_item} \
+            tax: {self.tax} tradable: {self.is_trade_accept} >"
+
+
+class DecisionEvent:
+    def __init__(
+        self, tick: int, type: FinanceType, item: int, sub_engine_name: str,
+        action_scope_func: Callable, action_type: ActionType
+    ):
+        """
+        Parameters:
+            tick (int): current tick of decision
+            type (FinanceType): type of this decision
+            item (int): available item index for action, such as a stock for StockSubEngine
+            sub_engine_name (str): name of sub-engine, used to identify which this decision come from,
+                as we support multi-sub-engine with same type, we need this field from action
+            action_scope_func (Callable): function to provide action scope
+        """
+        self.tick = tick
+        self.type = type
+        self.action_type = action_type
+        self.item = item
+        self.sub_engine_name = sub_engine_name
+        self._action_scope = None
+        self._action_scope_func = action_scope_func
+
+    @property
+    def action_scope(self) -> list:
+        """Action scope for items"""
+        if self._action_scope is None:
+            self._action_scope = self._action_scope_func(self.action_type, self.item)
+
+        return self._action_scope
+
+    def __repr__(self):
+        return f"<DecisionEvent type: {self.type}, tick: {self.tick}, engine: {self.sub_engine_name} >"
+
+
+class Action:
+    idx = 0
+
+    def __init__(
+        self, sub_engine_name: str, item_index: int = 0, number: int = 0,
+        action_type: ActionType = ActionType.order, tick: int = 0, order_mode: OrderMode = None,
+        stop: float = 0, limit: float = 0, id: int = None, life_time: int = 1
+    ):
+        """
+        Parameters:
+            sub_engine_name (str): name of engine the decision event from
+            item_index (int): index of the item (such as stock index), usually from DecisionEvent.items
+            number (int): number to perform, positive means buy, negitive means sell
+        """
+        self.sub_engine_name = sub_engine_name
+        self.item_index = item_index
+        self.number = number
+        self.action_type = action_type
```

Comment:
different action type has different attributes. Using a AbsAction and sub-Actions? 

We can have a discussion for this.

Reply:
done